### ISSUE TITLE
Add tests for optional chaining

### ISSUE THREAD

**Issue by DanielRosenwasser** [tc39/test262#2263]
Filing an issue to add tests for https://github.com/tc39/proposal-optional-chaining/

## Statement List

* ### UpdateExpression[Yield, Await]
  * - [ ] OptionalExpression[Yield, Await]
    * - [ ] MemberExpression[Yield, Await] **(member-expression.js)**
      * - [ ] PrimaryExpression[?Yield, ?Await]
        * - [ ] this
        * - [ ] IdentifierReference[?Yield, ?Await]
          * - [ ] Identifier
        * - [ ] Literal
        * - [ ] ArrayLiteral[?Yield, ?Await]
        * - [ ] ObjectLiteral[?Yield, ?Await]
        * - [ ] FunctionExpression
        * - [ ] ClassExpression[?Yield, ?Await]
        * - [ ] GeneratorExpression
        * - [ ] AsyncFunctionExpression
        * - [ ] AsyncGeneratorExpression
        * - [ ] RegularExpressionLiteral
        * - [ ] TemplateLiteral[?Yield, ?Await, ~Tagged]
        * - [ ] CoverParenthesizedExpressionAndArrowParameterList[?Yield, ?Await]
      * - [ ] MemberExpression[?Yield, ?Await][Expression[+In, ?Yield, ?Await]]
      * - [ ] MemberExpression[?Yield, ?Await].IdentifierName
      * - [ ] MemberExpression[?Yield, ?Await]TemplateLiteral[?Yield, ?Await, +Tagged]
      * - [ ] SuperProperty[?Yield, ?Await]
      * - [ ] MetaProperty
      * - [ ] newMemberExpression[?Yield, ?Await]Arguments[?Yield, ?Await]
    * - [ ] CallExpression[Yield, Await] **(call-expression.js)**
      * - [ ] CoverCallExpressionAndAsyncArrowHead[?Yield, ?Await]
      * - [ ] SuperCall[?Yield, ?Await]
      * - [ ] CallExpression[?Yield, ?Await]Arguments[?Yield, ?Await]
      * - [ ] CallExpression[?Yield, ?Await][Expression[+In, ?Yield, ?Await]]
      * - [ ] CallExpression[?Yield, ?Await].IdentifierName
      * - [ ] CallExpression[?Yield, ?Await]TemplateLiteral[?Yield, ?Await, +Tagged]
    * - [ ] OptionalExpression[Yield, Await]
      * - [ ] MemberExpression[?Yield, ?Await]OptionalChain[?Yield, ?Await]
      * - [ ] CallExpression[?Yield, ?Await]OptionalChain[?Yield, ?Await]
      * - [ ] OptionalExpression[?Yield, ?Await]OptionalChain[?Yield, ?Await]
  * - [ ] OptionalExpression[no LineTerminator here]++ (*invalid*) **(update-expression.js)**
  * - [ ] OptionalExpression[no LineTerminator here]-- (*invalid*) **(update-expression.js)**
  * - [ ] ++OptionalExpression[?Yield, ?Await] (*invalid*) **(update-expression.js)**
  * - [ ] --OptionalExpression[?Yield, ?]Await] (*invalid*) **(update-expression.js)**

* ### AssignmentExpression[In, Yield, Await] **(assignment-expression.js)**
  * - [ ] ConditionalExpression[?In, ?Yield, ?Await]
  * - [ ] ArrowFunction[?In, ?Yield, ?Await]
  * - [ ] AsyncArrowFunction[?In, ?Yield, ?Await]
  * - [ ] OptionalExpression[Yield?, Await?] = AssignmentExpression[?In, ?Yield, ?Await] (*invalid*)
  * - [ ] OptionalExpression AssignmentOperatorAssignmentExpression[?In, ?Yield, ?Await] (*invalid*)

* ### DestructuringAssignmentTarget[Yield, Await] **(assignment-expression.js)**
 * - [ ] OptionalExpression[?Yield, ?Await]

* ### IterationStatement[Yield, Await, Return]: **(iteration-statement.js)**
  * - [ ] doStatement[?Yield, ?Await, ?Return] while(Expression[+In, ?Yield, ?Await]);
  * - [ ] while(Expression[+In, ?Yield, ?Await]) Statement[?Yield, ?Await, ?Return]
  * - [ ] for([lookahead ∉ { let [ }]Expression[~In, ?Yield, ?Await]opt;Expression[+In, ?Yield, ?Await]opt;Expression[+In, ?Yield, ?Await]opt)Statement[?Yield, ?Await, ?Return]
  * - [ ] for(varVariableDeclarationList[~In, ?Yield, ?Await];Expression[+In, ?Yield, ?Await]opt;Expression[+In, ?Yield, ?Await]opt)Statement[?Yield, ?Await, ?Return]
  * - [ ] for(LexicalDeclaration[~In, ?Yield, ?Await]Expression[+In, ?Yield, ?Await]opt;Expression[+In, ?Yield, ?Await]opt)Statement[?Yield, ?Await, ?Return]
  * - [ ] for([lookahead ∉ { let [ }]LeftHandSideExpression[?Yield, ?Await]inExpression[+In, ?Yield, ?Await])Statement[?Yield, ?Await, ?Return]
  * - [ ] for(varForBinding[?Yield, ?Await]inExpression[+In, ?Yield, ?Await])Statement[?Yield, ?Await, ?Return]
  * - [ ] for(ForDeclaration[?Yield, ?Await]inExpression[+In, ?Yield, ?Await])Statement[?Yield, ?Await, ?Return]
  * - [ ] for([lookahead ≠ let]LeftHandSideExpression[?Yield, ?Await]ofAssignmentExpression[+In, ?Yield, ?Await])Statement[?Yield, ?Await, ?Return]
  * - [ ] for(varForBinding[?Yield, ?Await]ofAssignmentExpression[+In, ?Yield, ?Await])Statement[?Yield, ?Await, ?Return]
  * - [ ] for(ForDeclaration[?Yield, ?Await]ofAssignmentExpression[+In, ?Yield, ?Await])Statement[?Yield, ?Await, ?Return]

* ### ClassHeritage[Yield, Await]: **(class-heritage.js)**
  * - [ ] extends OptionalExpression[?Yield, ?Await]

## Async Function

* ### UpdateExpression[Yield, Await]
  * - [ ] OptionalExpression[Yield, Await]
    * - [ ] MemberExpression[Yield, Await] **(async-member-expression.js)**
      * - [ ] PrimaryExpression[?Yield, ?Await]
        * - [ ] this
        * - [ ] IdentifierReference[?Yield, ?Await]
          * - [ ] Identifier
          * - [ ] await Identifier
        * - [ ] Literal
        * - [ ] ArrayLiteral[?Yield, ?Await]
        * - [ ] ObjectLiteral[?Yield, ?Await]
        * - [ ] FunctionExpression
        * - [ ] ClassExpression[?Yield, ?Await]
        * - [ ] GeneratorExpression
        * - [ ] AsyncFunctionExpression
        * - [ ] AsyncGeneratorExpression
        * - [ ] RegularExpressionLiteral
        * - [ ] TemplateLiteral[?Yield, ?Await, ~Tagged]
        * - [ ] CoverParenthesizedExpressionAndArrowParameterList[?Yield, ?Await]
      * - [ ] MemberExpression[?Yield, ?Await][Expression[+In, ?Yield, ?Await]]
      * - [ ] MemberExpression[?Yield, ?Await].IdentifierName
      * - [ ] MemberExpression[?Yield, ?Await]TemplateLiteral[?Yield, ?Await, +Tagged]
      * - [ ] SuperProperty[?Yield, ?Await]
      * - [ ] MetaProperty
      * - [ ] newMemberExpression[?Yield, ?Await]Arguments[?Yield, ?Await]
    * - [ ] CallExpression[Yield, Await] **(async-call-expression.js)**
      * - [ ] CoverCallExpressionAndAsyncArrowHead[?Yield, ?Await]
      * - [ ] SuperCall[?Yield, ?Await]
      * - [ ] CallExpression[?Yield, ?Await]Arguments[?Yield, ?Await]
      * - [ ] CallExpression[?Yield, ?Await][Expression[+In, ?Yield, ?Await]]
      * - [ ] CallExpression[?Yield, ?Await].IdentifierName
      * - [ ] CallExpression[?Yield, ?Await]TemplateLiteral[?Yield, ?Await, +Tagged]
    * - [ ] OptionalExpression[Yield, Await]
      * - [ ] MemberExpression[?Yield, ?Await]OptionalChain[?Yield, ?Await]
      * - [ ] CallExpression[?Yield, ?Await]OptionalChain[?Yield, ?Await]
      * - [ ] OptionalExpression[?Yield, ?Await]OptionalChain[?Yield, ?Await]
  * - [ ] OptionalExpression[no LineTerminator here]++ (*invalid*) **(async-update-expression.js)**
  * - [ ] OptionalExpression[no LineTerminator here]-- (*invalid*) **(async-update-expression.js)**
  * - [ ] ++OptionalExpression[?Yield, ?Await] (*invalid*) **(async-update-expression.js)**
  * - [ ] --OptionalExpression[?Yield, ?]Await] (*invalid*) **(async-update-expression.js)**

* ### AssignmentExpression[In, Yield, Await] **(async-assignment-expression.js)**
  * - [ ] ConditionalExpression[?In, ?Yield, ?Await]
  * - [ ] ArrowFunction[?In, ?Yield, ?Await]
  * - [ ] AsyncArrowFunction[?In, ?Yield, ?Await]
  * - [ ] OptionalExpression[Yield?, Await?] = AssignmentExpression[?In, ?Yield, ?Await] (*invalid*)
  * - [ ] OptionalExpression AssignmentOperatorAssignmentExpression[?In, ?Yield, ?Await] (*invalid*)

* ### DestructuringAssignmentTarget[Yield, Await] **(async-assignment-expression.js)**
 * - [ ] OptionalExpression[?Yield, ?Await]

* ### IterationStatement[Yield, Await, Return]: **(async-iteration-statement.js)**
  * - [ ] doStatement[?Yield, ?Await, ?Return] while(Expression[+In, ?Yield, ?Await]);
  * - [ ] while(Expression[+In, ?Yield, ?Await]) Statement[?Yield, ?Await, ?Return]
  * - [ ] for([lookahead ∉ { let [ }]Expression[~In, ?Yield, ?Await]opt;Expression[+In, ?Yield, ?Await]opt;Expression[+In, ?Yield, ?Await]opt)Statement[?Yield, ?Await, ?Return]
  * - [ ] for(varVariableDeclarationList[~In, ?Yield, ?Await];Expression[+In, ?Yield, ?Await]opt;Expression[+In, ?Yield, ?Await]opt)Statement[?Yield, ?Await, ?Return]
  * - [ ] for(LexicalDeclaration[~In, ?Yield, ?Await]Expression[+In, ?Yield, ?Await]opt;Expression[+In, ?Yield, ?Await]opt)Statement[?Yield, ?Await, ?Return]
  * - [ ] for([lookahead ∉ { let [ }]LeftHandSideExpression[?Yield, ?Await]inExpression[+In, ?Yield, ?Await])Statement[?Yield, ?Await, ?Return]
  * - [ ] for(varForBinding[?Yield, ?Await]inExpression[+In, ?Yield, ?Await])Statement[?Yield, ?Await, ?Return]
  * - [ ] for(ForDeclaration[?Yield, ?Await]inExpression[+In, ?Yield, ?Await])Statement[?Yield, ?Await, ?Return]
  * - [ ] for([lookahead ≠ let]LeftHandSideExpression[?Yield, ?Await]ofAssignmentExpression[+In, ?Yield, ?Await])Statement[?Yield, ?Await, ?Return]
  * - [ ] for(varForBinding[?Yield, ?Await]ofAssignmentExpression[+In, ?Yield, ?Await])Statement[?Yield, ?Await, ?Return]
  * - [ ] for(ForDeclaration[?Yield, ?Await]ofAssignmentExpression[+In, ?Yield, ?Await])Statement[?Yield, ?Await, ?Return]
  * - [ ] [+Await]forawait([lookahead ≠ let]LeftHandSideExpression[?Yield, ?Await]ofAssignmentExpression[+In, ?Yield, ?Await])Statement[?Yield, ?Await, ?Return]
  * - [ ] [+Await]forawait(varForBinding[?Yield, ?Await]ofAssignmentExpression[+In, ?Yield, ?Await])Statement[?Yield, ?Await, ?Return]
  * - [ ] [+Await]forawait(ForDeclaration[?Yield, ?Await]ofAssignmentExpression[+In, ?Yield, ?Await])Statement[?Yield, ?Await, ?Return]

## Iterator Function

* ### UpdateExpression[Yield, Await]
  * - [ ] OptionalExpression[Yield, Await]
    * - [ ] MemberExpression[Yield, Await] **(iterator-member-expression.js)**
      * - [ ] PrimaryExpression[?Yield, ?Await]
        * - [ ] this
        * - [ ] IdentifierReference[?Yield, ?Await]
          * - [ ] yield Identifier

* ### AssignmentExpression[In, Yield, Await] **(iterator-assignment-expression.js)**
  * - [ ] [+Yield]YieldExpression[?In, ?Await]

**Comment by leobalter** [comment]
We already have an initial PR:

https://github.com/tc39/test262/pull/2212

**Comment by leobalter** [comment]
I updated the OT with the checklist contents produced by @bcoe.

**Comment by bcoe** [comment]
@leobalter @DanielRosenwasser if we're happy with this check list as a start, I'll put some open source time in this weekend and work my way through a bunch of the cases; I love this sort of work.

**Comment by DanielRosenwasser** [comment]
Only if you don't view it as an inconvenience - otherwise, I think it'd be awesome if you're up for it 😄 	

Leo will know more about the contribution process - ways that make it more approachable etc.

**Comment by leobalter** [comment]
This checklist is pretty comprehensive and I believe it's a great start. We can add interesting edgy cases along the way as we identify them.

BTW: one edgy case that I discussed with my team this week, based on #2262:

```js
var stored = [];
var x = {
  set y(v) { stored.push(`set y with ${v}`); }
  get y() { stored.push('get y'); }
};
// dstr assignment
[ x?.y = 1 ] = [ 42 ];
assert.sameValue(stored.length, 1);
assert.sameValue(stored[0], 42);
```

**Comment by ljharb** [comment]
Im confused; why would optional chaining be permitted in a destructuring context?

**Comment by leobalter** [comment]
Checkout the ref issue. It’s valid to have a LHSExpr inside an
ArrayAssignment literal. Which I believe makes it consistent to have opt
chaining there.

**Comment by leobalter** [comment]
Well, https://tc39.es/proposal-optional-chaining/#sec-static-semantics-static-semantics-isvalidsimpleassignmenttarget

There is an explicit exception for this grammar, so `[x.y = 1] = [42]`  remains valid while `[x?.y = 1] = [42]` is a parse error. We still need test for that, this is a documented and expected error.

**Comment by bcoe** [comment]
it would be good to add tests for the edge-cases demonstrated here:

https://github.com/v8/v8/commit/ceb7bd59432a340ea3711ef5acb8c148471eee8c#diff-9b0c860cbf7e0482f43ad30124ee9f9bR92

Mainly:

* `super?.foo` calls.
* `new?.bar` calls.

**Comment by bcoe** [comment]
CC: @devsnek :wave: saw you've already done some initial implementation work in V8, awesome :smile: 

we're gradually fleshing out a thorough set of tests [here](https://github.com/tc39/test262/pull/2272/files) which you can run against, let me know if you bump into any issues.

**Comment by bcoe** [comment]
Just a quick update, I have started work on the async tests for optional chaining (have just been fairly busy with a few other tasks).

**Comment by bcoe** [comment]
@DanielRosenwasser @leobalter @rkirsling, it feels like we could still improve coverage for Stage 4 in December. I saw the rollback of #2411, where should we ultimately end up in terms of generated tests vs., hand written tests, is it worth continuing to work down the [check-lists we've written up](https://gist.github.com/bcoe/b0e6ae3f93f8c8b8f8559f4428187f25) ... alternatively, perhaps we're feeling that we've covered a healthy number of productions?

**Comment by leobalter** [comment]
@bcoe would you be able (time-wise) to update the coverage list or flag parts missing in the coverage?

We can make a work effort after that or as you said, define coverage is reasonable enough. WDYT?

**Comment by bcoe** [comment]
@leobalter I'll make an effort to put some work into it this weekend, at least summarizing some of the productions we didn't quite hit yet.

I definitely do think we're getting to the point where we have a healthy number of tests, but I think there are some areas we could flesh out more (as an example for and while statements).

**Comment by bcoe** [comment]
@leobalter @jridgewell @DanielRosenwasser, I'm feeling pretty good about our initial suite of tests (once we land #2429 which addresses some regressions @jridgewell brought up).
